klippy/extras/sht4x.py

```diff
@@ -0,0 +1,158 @@
+RESET = 0x94
+
+# Measurement command (high precision)
+MEASUREMENT_CMD = 0xFD
+
+# Measurement timing constant
+MEASUREMENT_TIME = 0.009  # 9ms
+RESET_TIME = 0.001  # 1ms
+
+class SHT4X:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.name = config.get_name().split()[-1]
+        self.reactor = self.printer.get_reactor()
+
+        # I2C setup
+        addr = config.getint('i2c_address', DEFAULT_ADDR)
+        self.i2c = bus.MCU_I2C_from_config(config, default_addr=addr)
+        self.mcu = self.i2c.get_mcu()
+
+        # Sensor state
+        self.temp = 0
+        self.humidity = 0
+        self._error_count = 0
+        self.sensor_ready = False
+
+        # Register with Klipper
+        self.printer.add_object("sht4x " + self.name, self)
+        self.printer.register_event_handler(
+            "klippy:connect", self.handle_connect)
+
+    def handle_connect(self):
+        """Initialize sensor"""
+        try:
+            self.reset()
+            if self._test_measurement():
+                self.sensor_ready = True
+                logging.info("SHT4X: Sensor ready")
+                self.sample_timer = self.reactor.register_timer(
+                    self.sample_sensor)
+                self.reactor.update_timer(self.sample_timer, self.reactor.NOW)
+            else:
+                logging.error("SHT4X: Sensor test failed")
+        except Exception as e:
+            logging.error("SHT4X: Init failed: %s" % str(e))
+
+    def _test_measurement(self):
+        """Quick measurement test"""
+        try:
+            recv = self.get_measurements()
+            return self._validate_crc(recv)
+        except Exception:
+            return False
+
+    def _validate_crc(self, data):
+        """Validate CRC for both temp and humidity"""
+        temp_crc = data[2]
+        humidity_crc = data[5]
+        return (temp_crc == self._crc8(data[0:2]) and
+                humidity_crc == self._crc8(data[3:5]))
+
+    def setup_minmax(self, min_temp, max_temp):
+        pass  # Required by Klipper interface
+
+    def setup_callback(self, cb):
+        self._callback = cb
+
+    def sample_sensor(self, eventtime):
+        """Main sensor sampling method"""
+        if not self.sensor_ready:
+            return eventtime + 1.0
+
+        try:
+            # Get measurements
+            recv = self.get_measurements()
+
+            if not self._validate_crc(recv):
+                raise Exception("CRC validation failed")
+
+            # Process temperature
+            raw_temp = (recv[0] << 8) | recv[1]
+            self.temp = -45.0 + 175.0 * raw_temp / 65535.0
+
+            # Process humidity
+            raw_humidity = (recv[3] << 8) | recv[4]
+            humidity_percent = -6.0 + 125.0 * raw_humidity / 65535.0
+            self.humidity = max(min(humidity_percent, 100.0), 0.0)
+
+            self._error_count = 0
+
+            # Report to Klipper
+            if hasattr(self, '_callback'):
```

Comment:
A resourceful solution. Alas, as far as I'm aware, there is always some callback.
Generally, we do want all used attributes to be defined in the init section.